File: JS_module1/threesum.js
arr = [12, 3, 1, 2, -6, 5, -8, 6]
target = 0

const threesum = (array, targetSum) => {
    array.sort((a, b) => a - b)
    const triplets = [];
    for (var i = 0; i < array.length - 2; i++) {
        let left = i + 1;
        let right = array.length - 1
        while (left < right) {
            currentSum = array[i] + array[left] + array[right]
            if (currentSum == targetSum) {
                triplets.push([array[i], array[left], array[right]])
                left++;
                right--;
            } else if (currentSum < targetSum) {
                left++;
            } else if (currentSum > targetSum) {
                right--;
            }
        }
    }
    return triplets
}

console.log(threesum(arr, target))

File: PY_module1/AlgoExpert_Medium/threesum.py
arr = [12, 3, 1, 2, -6, 5, -8, 6]
target = 0

# concepts in this function:
# sort, => array.sort()
# create an empty array
# loop through array except the first digit, => for i in range(len(array) - 2)
## assign the left pointer to be the 2nd digit, => left = i + 1
## assign the right pointer to be the far most digit => len(array) - 1
## while loop left pointer less than right pointer
### define currentSum = 3 numbers in the array => use indexes
### if currentSum == target, append the 3 numbers in an array
### increment the left pointer
### decrement the right pointer
### if currentSum < target, increment left pointer
### if currentSum > target, decerement right pointer
# return the array

def threesum(array, targetSum):
    array.sort()
    triplets = []
    for i in range(len(array) - 2):
        left = i + 1
        right = len(array) - 1
        while left < right:
            currentSum = array[i] + array[left] + array[right]
            if currentSum == targetSum:
                triplets.append([array[i], array[left], array[right]])
                left += 1
                right -= 1
            elif currentSum < targetSum:
                left += 1
            elif currentSum > targetSum:
                right -= 1
    return triplets

print(threesum(arr, target))